sdcm/nemesis.py

```diff
@@ -4417,6 +4423,106 @@ def disrupt_add_remove_mv(self):
 
+    def disrupt_toggle_audit(self, store: AuditStore):
+        """
+            Enable audit log with all categories and user keyspaces (if audit already enabled, disable it and finish the Nemesis),
+            verify audit log content,
+            reduce categories by excluding DML and QUERY,
+            verify DDL are logged in audit log correctly. Leaves audit log enabled this way.
+        """
+        if not self.target_node.is_enterprise:
+            raise UnsupportedNemesis("Auditing feature is only supported by Scylla Enterprise")
+
+        if store == "syslog" and self._is_it_on_kubernetes():
+            raise UnsupportedNemesis("syslog store is not supported on Kubernetes #scylla-operator/1299")
+
+        audit = Audit(self.cluster)
+
+        if audit.is_enabled():
+            audit.disable()
+            raise UnsupportedNemesis("Audit was enabled -> disabling it")
+
+        audit_keyspace = "audit_keyspace"
+        keyspaces_for_audit = [audit_keyspace]
+        InfoEvent(f"Enabling full audit for keyspaces: {keyspaces_for_audit}").publish()
+        audit_config = AuditConfiguration(
+            store=store,
+            categories=["DCL", "DDL", "AUTH", "ADMIN", "DML", "QUERY"],
+            keyspaces=keyspaces_for_audit,
+            tables=[],
+        )
+        try:
+            audit.configure(audit_config)
+            keyspace_name = keyspaces_for_audit[0]
+            errors = []
+            audit_start = datetime.datetime.now()
+            InfoEvent(message='Writing/Reading data from audited keyspace').publish()
+            write_cmd = f"cassandra-stress write no-warmup cl=ALL n=1000 -schema" \
+                        f" 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)" \
+                        f" keyspace={audit_keyspace}' -mode cql3 native -rate 'threads=1 throttle=1000/s'" \
+                        f" -pop seq=1..1000 -col 'n=FIXED(1) size=FIXED(128)' -log interval=5"
+            write_thread = self.tester.run_stress_thread(
+                stress_cmd=write_cmd, round_robin=True, stop_test_on_failure=False)
+            self.tester.verify_stress_thread(cs_thread_pool=write_thread)
+            read_cmd = f"cassandra-stress read no-warmup cl=ALL n=1000 " \
+                       f" -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)" \
+                       f" keyspace={audit_keyspace}' -mode cql3 native -rate 'threads=1 throttle=1000/s'" \
+                       f" -pop seq=1..1000 -col 'n=FIXED(1) size=FIXED(128)' -log interval=5"
+            read_thread = self.tester.run_stress_thread(
+                stress_cmd=read_cmd, round_robin=True, stop_test_on_failure=False)
+            self.tester.verify_stress_thread(cs_thread_pool=read_thread)
+            InfoEvent(message='Verifying Audit table contents').publish()
+            rows = audit.get_audit_log(from_datetime=audit_start, category="DML", limit_rows=1100)
```

Comment:
I assume that we begin with a clear `audit.audit_log`, right? And we're fetching 1.1k rows, while inserting only 1k, just to make sure we really fetched all of the rows inserted in this session. But what if, by mistake, e.g. 1.2k rows have been inserted, 200 coming from `USE KS`. We'd remove these 200 extra rows, but because we only fetched 1.1 rows, we'd be left with 1100-200=900 rows, and 900 != 1000, so the next check would fail, but perhaps there's really 1k DML rows in audit, it's just that we haven't fetched all of them. I guess that the bottom line here is that we should maybe not limit the number of rows fetched.
OTOH, if `audit.audit_log` is not empty, we'd be fetching garbage from previous runs? - That's why assume it is empty :)

Reply:
Also, maybe we should log the number of rows fetched for debugging purposes.

Reply:
so, the idea to overcome these issues is to first fetch only ones that are after `from_datetime` - which is right before c-s start. The second, you are right in case of 200 `USE KS` statements we might miss valid rows, possibly we can drop it. From the other hand, assuming it is working now, when we start to see 200 `USE KS` won't it be some kind of regression than suddenly we started to log so many of these?

Reply:
> Also, maybe we should log the number of rows fetched for debugging purposes.

we do it in `get_audit_log` method